nightly-examples/examples/proxy_server.rs

```diff
@@ -0,0 +1,130 @@
+#![deny(rust_2018_idioms)]
+
+//! A proxy that forwards data to another server and forwards that server's
+//! responses back to clients.
+//!
+//! Because the Tokio runtime uses a thread pool, each TCP connection is
+//! processed concurrently with all other TCP connections across multiple
+//! threads.
+//!
+//! You can showcase this by running this in one terminal:
+//!
+//!     cargo +nightly run --example proxy_server
+//!
+//! This in another terminal
+//!
+//!     cargo +nightly run --example echo
+//!
+//! And finally this in another terminal
+//!
+//!     nc localhost 8081
+//!
+//! This final terminal will connect to our proxy, which will in turn connect to
+//! the echo server, and you'll be able to see data flowing between them.
+
+use futures::{future::try_join, TryFutureExt};
+use tokio::{
+    self,
+    io::AsyncReadExt,
+    net::{TcpListener, TcpStream},
+    prelude::*,
+};
+
+use tracing::{debug, debug_span, info, warn};
+use tracing_attributes::instrument;
+use tracing_futures::Instrument;
+
+use std::{env, net::SocketAddr};
+
+#[instrument]
+async fn transfer(
+    inbound: TcpStream,
+    proxy_addr: SocketAddr,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let outbound = TcpStream::connect(&proxy_addr).await?;
+
+    let (mut ri, mut wi) = inbound.split();
+    let (mut ro, mut wo) = outbound.split();
+
+    let client_to_server = ri
+        .copy(&mut wo)
+        .map(|bytes| {
+            if let Ok(n) = bytes {
+                debug!(message = "copy finished", n);
+            }
+
+            bytes
+        })
+        .map_err(|error| {
+            warn!(%error);
+            error
+        })
+        .instrument(debug_span!("client_to_server"));
+    let server_to_client = ro
+        .copy(&mut wi)
+        .map(|bytes| {
+            if let Ok(n) = bytes {
+                debug!(message = "copy finished", n);
+            }
+
+            bytes
+        })
+        .map_err(|error| {
+            warn!(%error);
```

Comment:
And, as above, if we use `tracing` 0.1.6, we can say:
```suggestion
            warn!(error);
```